x/supplier/keeper/query_supplier.go

```diff
@@ -31,11 +32,23 @@ func (k Keeper) AllSuppliers(
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 
-	if req.GetServiceId() == "" {
-		return k.getAllSuppliers(ctx, logger, req)
-	} else {
+	// Determine the most efficient query strategy based on filters
+	hasServiceId := req.GetServiceId() != ""
+	hasOperatorAddress := req.GetOperatorAddress() != ""
+	hasOwnerAddress := req.GetOwnerAddress() != ""
+
+	// Strategy 1: Direct lookup by operator_address (most efficient)
+	if hasOperatorAddress && !hasServiceId && !hasOwnerAddress {
+		return k.getSupplierByOperator(ctx, logger, req)
+	}
+
+	// Strategy 2: Filter by service_id (using service index)
+	if hasServiceId {
 		return k.getAllServiceSuppliers(ctx, logger, req)
```

Comment:
If `getSupplierByOperator` handles all queries that have `operator` as filter (as mentioned above).

This `getAllServiceSuppliers` could handle any query that have `service` or `service AND owner` as filters.

It would use the `service` index as an initial filter, and apply the `operator` check at each iteration.